lsbatch/man5/lsb.params.5
.ds ]W %
.ds ]L
.nh
.TH lsb.params 5 "OpenLava Version 4.0 - Aug 2016"
.br
.SH NAME
\fBlsb.params\fR
.SS \fB\fROverview
.BR
The lsb.params file defines general parameters used by openlava Batch.
This file contains only one section, named Parameters. MBD uses
lsb.params for initialization. The file is optional. If not present, the
openlava-defined defaults are assumed.
.PP
Some of the parameters that can be defined in lsb.params control
timing within the openlava Batch system. The default settings provide good
throughput for long-running batch jobs while adding a minimum of
processing overhead in the batch daemons.
.SS Contents
.BR
.RS
.HP 2
\(bu Parameters Section
.RE
.SH Parameters Section
.BR
This section and all the keywords in this section are optional. If
keywords are not present, openlava Batch assumes default values for the
corresponding keywords. The valid keywords for this section are:
.SH ABS_RUNLIMIT
.BR
.PP
.SS Syntax
.BR
\fBABS_RUNLIMIT = y | Y\fR
.SS Description
.BR
Specify y or Y to use the absolute run time for controlling the job run limit.
.SS Default
.BR
Undefined. The normalized job run time is used for the job run limit.
.SH ACCT_ARCHIVE_AGE
.BR
.PP
.SS Syntax
.BR
\fBACCT_ARCHIVE_AGE = \fR\fIdays\fR
.SS Description
.BR
Enables automatic archiving of openlava accounting log files, and specifies
the archive interval. openlava archives the current log file if the length of
time from its creation date exceeds the specified number of days.
.SS See Also
.BR
ACCT_ARCHIVE_SIZE also enables automatic archiving.
.PP
MAX_ACCT_ARCHIVE_FILE enables automatic deletion of the
archives.
.SS Default
.BR
.PP
.PP
Undefined (no limit to the age of lsb.acct).
.SH ACCT_ARCHIVE_SIZE
.BR
.SS Syntax
.BR
\fBACCT_ARCHIVE_SIZE = \fR\fIkilobytes\fR
.SS Description
.BR
Enables automatic archiving of openlava accounting log files, and specifies
the archive threshold. openlava archives the current log file if its size exceeds
the specified number of kilobytes.
.SS See Also
.BR
ACCT_ARCHIVE_AGE also enables automatic archiving.
.PP
MAX_ACCT_ARCHIVE_FILE enables automatic deletion of the
archives.
.SS Default
.BR
Undefined (no limit to the size of lsb.acct).
.SH CLEAN_PERIOD
.BR
.SS Syntax
.BR
\fBCLEAN_PERIOD\fR \fB=\fR \fIseconds\fR
.SS Description
.BR
For non-repetitive jobs, the amount of time that job records for jobs that
have finished or have been killed are kept in MBD core memory after
they have finished.
.PP
Users can still see all jobs after they have finished using the \fBbjobs\fR
command. For jobs that finished more than CLEAN_PERIOD seconds
ago, use the \fBbhist\fR command.
.SS Default
.BR
3600 (1 hour).
.SH DEFAULT_HOST_SPEC
.BR
.SS Syntax
.BR
\fBDEFAULT_HOST_SPEC =\fR \fIhost_name | host_model\fR
.SS Description
.BR
The default CPU time normalization host for the cluster.
.PP
The CPU factor of the specified host or host model will be used to
normalize the CPU time limit of all jobs in the cluster, unless the CPU
time normalization host is specified at the queue or job level.
.SS Default
.BR
Undefined
.SH DEFAULT_PROJECT
.BR
.SS Syntax
.BR
\fBDEFAULT_PROJECT\fR \fB=\fR \fIproject_name\fR
.SS Description
.BR
The name of the default project. Specify any string.
.PP
When you submit a job without specifying any project name, and the
environment variable LSB_DEFAULTPROJECT is not set, openlava
automatically assigns the job to this project.
.SS Default
.BR
default
.SH DEFAULT_QUEUE
.BR
.SS Syntax
.BR
\fBDEFAULT_QUEUE\fR \fB=\fR \fIqueue_name \fR...
.SS Description
.BR
Space-separated list of candidate default queues (candidates must
already be defined in lsb.queues).
.PP
When you submit a job to openlava without explicitly specifying a queue,
and the environment variable LSB_DEFAULTQUEUE is not set, openlava puts
the job in the first queue in this list that satisfies the job's specifications
subject to other restrictions, such as requested hosts, queue status, etc.
.SS Default
.BR
Undefined. When a user submits a job to openlava without explicitly
specifying a queue, and there are no candidate default queues defined
(by this parameter or by the user's environment variable
LSB_DEFAULTQUEUE), openlava automatically creates a new queue named
default, using the default configuration, and submits the job to that
queue.
.SH DISABLE_UACCT_MAP
.BR
.SS Syntax
.BR
\fBDISABLE_UACCT_MAP = y | Y\fR
.SS Description
.BR
Specify y or Y to disable user-level account mapping.
.SS Default
.BR
Undefined
.SH  HIST_MINUTES
.BR
The number of minutes the fairshare schedule performs the decay of accumulated
number of used slots per user in a fairshare tree. The number of running slots
is decreased ten times each \fBHIST_MINUTES\fR. This number is used to sort the
tree by priority, smaller the number less important is the history os used slots.
The default value is two hours.
.BR
.SH JOB_ACCEPT_INTERVAL
.BR
.SS Syntax
.BR
\fBJOB_ACCEPT_INTERVAL =\fR \fIinteger\fR
.SS Description
.BR
The number of dispatch turns to wait after dispatching a job to a host,
before dispatching a second job to the same host. By default, a dispatch
turn lasts 60 seconds (MBD_SLEEP_TIME in lsb.params).
.PP
If 0 (zero), a host may accept more than one job in each job dispatching
interval. By default, there is no limit to the total number of jobs that can
run on a host, so if this parameter is set to 0, a very large number of
jobs might be dispatched to a host all at once. You may notice
performance problems if this occurs.
.PP
JOB_ACCEPT_INTERVAL set at the queue level (lsb.queues)
overrides JOB_ACCEPT_INTERVAL set at the cluster level
(lsb.params).
.SS Default
.BR
1
.SH JOB_DEP_LAST_SUB
.BR
.SS Description
.BR
Used only with job dependency scheduling.
.PP
If set to 1, whenever dependency conditions use a job name that
belongs to multiple jobs, openlava evaluates only the most recently
submitted job.
.PP
Otherwise, all the jobs with the specified name must satisfy the
dependency condition.
.SS Default
.BR
Undefined
.SH JOB_PRIORITY_OVER_TIME
.BR
.SS Syntax
.BR
\fBJOB_PRIORITY_OVER_TIME=\fR\fIincrement\fR\fB/\fR\fIinterval
\fR
.SS Description
.BR
JOB_PRIORITY_OVER_TIME enables automatic job priority escalation
when MAX_USER_PRIORITY is also defined.
.SS Valid Values
.BR
\fIincrement\fR
.IP
Specifies the value used to increase job priority every \fIinterval\fR
minutes. Valid values are positive integers.

.RE
.PP
\fIinterval\fR
.IP
Specifies the frequency, in minutes, to \fIincrement\fR job priority.
Valid values are positive integers.

.RE
.SS Default
.BR
Undefined
.SS Example
.BR
JOB_PRIORITY_OVER_TIME=3/20
.IP
Specifies that every 20 minute \fIinterval\fR \fIincrement\fR to job priority
of pending jobs by 3.

.RE
.SS See Also
.BR
MAX_USER_PRIORITY
.SH JOB_SPOOL_DIR
.BR
.SS Syntax
.BR

\fBJOB_SPOOL_DIR =\fR \fIdir\fR
.SS Description
.BR

Specifies the directory for buffering batch standard output and standard
error for a job
.PP
When JOB_SPOOL_DIR is defined, the standard output and standard
error for the job is buffered in the specified directory.
.PP
Except for \fBbsub -is\fR and \fBbsub -Zs\fR, if JOB_SPOOL_DIR is not
accessible or does not exist, output is spooled to the default job output
directory .lsbatch.
.PP
For \fBbsub -is\fR and \fBbsub -Zs\fR, JOB_SPOOL_DIR must be readable and
writable by the job submission user, and it must be shared by the
master host, the submission host, and the execution host. If the
specified directory is not accessible or does not exist, \fBbsub -is\fR and
\fBbsub -Zs\fR cannot write to the default directory and the job will fail.
.PP
As openlava runs jobs, it creates temporary directories and files under
JOB_SPOOL_DIR. By default, openlava removes these directories and files
after the job is finished. See \fBbsub\fR(\fB1\fR) for information about job
submission options that specify the disposition of these files.
.PP
On UNIX, specify an absolute path. For example:

.PP
JOB_SPOOL_DIR=/home/share/lsf_spool

.PP
JOB_SPOOL_DIR can be any valid path up to a
maximum length of 256 characters. This maximum path length includes
the temporary directories and files that openlava Batch creates as jobs run.
The path you specify for JOB_SPOOL_DIR should be as short as
possible to avoid exceeding this limit.
.SS Default
.BR
Undefined
.PP
Batch job output (standard output and standard error) is sent to the
.lsbatch directory on the execution host:
.RS
.HP 2
\(bu On UNIX: $HOME/.lsbatch
.RE

.IP
If %HOME% is specified in the user environment, uses that
directory instead of %windir% for spooled output.
.SH JOB_TERMINATE_INTERVAL
.BR
.SS Syntax
.BR

\fBJOB_TERMINATE_INTERVAL =\fR \fIseconds\fR
.SS Description
.BR

Specifies the time interval in seconds between sending SIGINT,
SIGTERM, and SIGKILL when terminating a job. When a job is
terminated, the job is sent SIGINT, SIGTERM, and SIGKILL in sequence
with a sleep time of JOB_TERMINATE_INTERVAL between sending the
signals. This allows the job to clean up if necessary.
.SS Default
10
.SH LOAD_UPDATE_INTVL
.BR
.SS Syntax
.BR
LOAD_UPDATE_INTVL=\fIseconds\fR
.SS Description
.PP
How often the MBD calls LIM to update the load and shared resource
information. Higher the interval and more sentive to load and resource
information changes the scheduler is.
.SS Default
15 seconds
.SH MAX_ACCT_ARCHIVE_FILE
.BR
.SS Syntax
.BR
MAX_ACCT_ARCHIVE_FILE=\fIinteger\fR
.SS Description
.BR

Enables automatic deletion of archived openlava accounting log files and
specifies the archive limit.
.SS Compatibility
.BR

ACCT_ARCHIVE_SIZE or ACCT_ARCHIVE_AGE should also be
defined.
.SS Example
.BR
MAX_ACCT_ARCHIVE_FILE=10
.PP
openlava maintains the current lsb.acct and up to 10 archives. Every time
the old lsb.acct.9 becomes lsb.acct.10, the old lsb.acct.10
gets deleted.
.SS Default
.BR
Undefined (no deletion of lsb.acct.\fIn\fR files).
.SH MAX_JOB_ARRAY_SIZE
.BR
.SS Syntax
.BR
\fBMAX_JOB_ARRAY_SIZE =\fR \fIinteger\fR
.SS Description
.BR
Specifies the maximum index value of a job array that can be created
by a user for a single job submission. The maximum number of jobs in
a job array cannot exceed this value, and will be less if some index
values are not used (start, end, and step values can all be used to limit
the indices used in a job array).
.PP
A large job array allows a user to submit a large number of jobs to the
system with a single job submission.
.PP
Specify an integer value from 1 to 65534.
.SS Default
.BR
1000
.SH MAX_JOBID
.BR
.SS Syntax
.BR
\fBMAX_JOBID=\fR\fIinteger\fR
.SS Description
.BR
The job ID limit. The job ID limit is the highest job ID that openlava will ever
assign, and also the maximum number of jobs in the system.
.PP
Specify any integer from 999999 to 9999999 (for practical purposes, any
seven-digit integer).
.SS Example
.BR
MAX_JOBID=1234567
.SS Default
.BR
999999
.SH MAX_JOB_NUM
.BR
.SS Syntax
.BR

\fBMAX_JOB_NUM\fR \fB=\fR \fIinteger\fR
.SS Description
.BR
The maximum number of finished jobs whose events are to be stored
in the lsb.events log file.
.PP
Once the limit is reached, MBD starts a new event log file. The old
event log file is saved as lsb.events.\fIn\fR, with subsequent sequence
number suffixes incremented by 1 each time a new log file is started.
Event logging continues in the new lsb.events file.
.SS Default
.BR

1000
.SH MAX_SBD_FAIL
.BR

.SS Syntax
.BR

\fBMAX_SBD_FAIL = \fR\fIinteger\fR
.SS Description
.BR

The maximum number of retries for reaching a non-responding slave
batch daemon, SBD.
.PP
The interval between retries is defined by MBD_SLEEP_TIME. If MBD
fails to reach a host and has retried MAX_SBD_FAIL times, the host is
considered unavailable. When a host becomes unavailable, MBD
assumes that all jobs running on that host have exited and that all
rerunnable jobs (jobs submitted with the \fBbsub\fR \fB-r\fR option) are
scheduled to be rerun on another host.
.SS Default
.BR
3
.SH MAX_SBD_CONNS
.BR
.PP
.SS Syntax
.BR
\fBMAX_SBD_CONNS = \fR\fIinteger\fR
.SS Description
.BR
The maximum number of files mbatchd can have open and connected
to sbatchd
.SH MAX_SCHED_STAY
.BR
.PP
.SS Syntax
.BR
\fBMAX_SCHED_STAY = \fR\fIinteger\fR
.SS Description
.BR
The time in seconds the mbatchd has for scheduling pass.
.SS Default
.BR
3
.SH MAX_STREAM_RECORDS
.BR
.SS Syntax
.BR
\fBMAX_STREAM_RECORDS=\fR\fIinteger\fR
.SS Description
.BR
Enables mbatchd to write job events data into a stream file
lsb.stream.0 located
in $LSB_SHAREDIR/logdir/stream. This is to allow a 3rd party
program to read and process job data in real time
without touching lsb.events files.

When mbatchd switches lsb.events file (i.e. renames the current event
file to lsb.events.1 and creates a new lsb.events file),
the 3rd party program is not informed for the change. As the result,
it is hard for the 3rd party program to rely on lsb.events file for
streaming job data. In addition, lsb.events files are critical in
the OpenLava system. It is not recommended for a 3rd party program
to touch them.

After writing number of records that is defined by
MAX_STREAM_RECORDS, the mbatchd writes an STREAM_END event,
closes the file lsb.stream.0, renames it to lsb.stream.1, and
creates a new lsb.stream.0 to conitnue writing job events.
The previous lsb.stream.1 is overridden.

The 3rd party program could call OpenLava C API lsb_geteventrec() to
read and parse records in lsb.stream.0.
The program should close the file after getting the event
STREAM_END, then open the file lsb.stream.0 again to continue reading.
.PP
.SS Default
.BR
Undefined. No stream files are created.
.SS See Also
lsb.events(5) lsb_geteventrec(3)
.SH MAX_USER_PRIORITY
.BR
.SS Syntax
.BR
\fBMAX_USER_PRIORITY=\fR\fIinteger\fR
.SS Description
.BR
Enables user-assigned job priority and specifies the maximum job
priority a user can assign to a job.
.PP
openlava administrators can assign a job priority higher than the specified
value.
.SS Compatibility
.BR
User-assigned job priority changes the behavior of \fBbtop\fR and \fBbbot\fR.
.SS Example
.BR
MAX_USER_PRIORITY=2.0
.PP
Specifies that 2.0 is the maximum job priority that can be specified by
a user.
.SS Default
.BR
Undefined
.SS See Also
.BR
bsub, bmod, btop, bbot, JOB_PRIORITY_OVER_TIME
.SH MBD_SLEEP_TIME
.BR
.SS Syntax
.BR
\fBMBD_SLEEP_TIME =\fR \fIseconds\fR
.SS Description
.BR
The job dispatching interval; how often openlava tries to dispatch pending
jobs.
.SS Default
.BR
60
.SH PG_SUSP_IT
.BR
.SS Syntax
.BR
\fBPG_SUSP_IT =\fR \fIseconds\fR
.SS Description
.BR
The time interval that a host should be interactively idle (it > 0) before
jobs suspended because of a threshold on the pg load index can be
resumed.
.PP
This parameter is used to prevent the case in which a batch job is
suspended and resumed too often as it raises the paging rate while
running and lowers it while suspended. If you are not concerned with
the interference with interactive jobs caused by paging, the value of
this parameter may be set to 0.
.SS Default
.BR
2.0 (seconds)
.SH PREEMPTABLE_RESOURCES
.BR
.SS Syntax
.BR
\fBPREEMPTABLE_RESOURCES =\fR "\fIresource ...\fR"
.SS Description
.PP
Configuration to enable resource based job preemption. Only resources defined in
the resource name list (names separated by spaces)
can trigger the job preemption.
.PP
If this parameter is not configured, the preemption is based on job slots.
Enabling the resource based job preemption disables the slot based preemption.
.PP
When a job in a high priority queue is pending because its required resource
is unavailable, and the resource is defined in PREEMPTABLE_RESOURCES,
OpenLava will preempt low priority jobs that are using the resource, until
the resource is available for the high priority job to run.
This is often used for application license preemption.
.PP
To enable the preemptioni policy, the parameter PREEMPTION needs to be set in the
high priority queue configuration. Please see lsb.queues(5) for the detail.
.PP
The resource based job preemption is only triggered when the high priority job
required resource is not available, the resource is used by low priority jobs,
and there are free job slots. If there are no free job slots, resource based
preemption is not triggered.
.SS Default
Undefined. Resource based preemption is disabled. Preemption is based on job
slots.
.SH PREEMPT_SLOT_SUSPEND
.BR
.SS Syntax
.BR
\fBPREEMPT_SLOT_SUSPEND =\fR y | n
.SS Description
.BR
Specifies the behavior of the slot based preemption. 
If \fBPREEMPT_SLOT_SUSPEND\fR is set to "y", the job preemption 
suspends the preempted job. If it is set to "n",
the job preemption kills then requeues the preempted job.
.SS Default
.BR
n
.SH SBD_SLEEP_TIME
.BR
.SS Syntax
.BR
\fBSBD_SLEEP_TIME =\fR \fIseconds\fR
.SS Description
.BR
The interval at which openlava checks the load conditions of each host, to
decide whether jobs on the host must be suspended or resumed.
.SS Default
.BR
30
.SH SHARED_RESOURCE_UPDATE_FACTOR
.BR
.SS Syntax
.BR
\fBSHARED_RESOURCE_UPDATE_FACTOR = \fR\fIinteger\fR
.SS Description
.BR
Determines the static shared resource update interval for the cluster.
.PP
Specify approximately how many times to update static shared
resources during one MBD sleep time period. The formula is:
.PP
\fIinterval\fR = MBD_SLEEP_TIME /
SHARED_RESOURCE_UPDATE_FACTOR
.PP
where the result of the calculation is truncated to an integer. The static
shared resource update interval is in seconds.
.SS Default
.BR
Undefined (all resources are updated only once, at the start of each
dispatch turn).


NEWS
Release notes for OpenLava versions

Master
========

-- Implement SLOTS_PER_PROCESSOR in lsb.resources.

4.0beta2
============

-- Add Platform MPI wrapper.
-- Use bash for etc/openlava to support Ubuntu.
-- Fixed rusage string parsing bug.
-- Introduce USUSP and SSUSP counters in the fairshare accounts. Previously
   suspended jobs were accounted as running leading to erroneous state
   transition from suspended to pending state.
-- Implement a new lsf.conf variable named LSB_DEFAULT_USER_GROUP="group name".
   If this variable is specified than jobs that do not specify -G at submission
   time will be assigned LSB_DEFAULT_USER_GROUP as group. This group must be
   configured among valid groups in a fairshare queue.
-- In the fairshare scheduling if a user has jobs that enforce a user group
   and jobs that don't, schedule both types of jobs with -G jobs first.
-- Implement a set of fixes related to fairshare job selection which were
   causing the scheduler to slowly dispatch jobs even if slots were available
   and set incomplete pending reasons.
-- Fixed slot based and resource preemption job replay bug. Fixed slot based
   preemption memory leak, and man pages.
-- Implement a slot based queue preemption in which jobs are suspended,
   instead of requeued.
-- If the MBD does not set the pending reasons for a job print:
   "System is unable to set pending reason for job;" in the bjobs -p
   output instead of blank or meaningless string.
-- badmin ckconfig checks the ownership and permission of work/logdir.
-- In bqueues -l output change the keyword RAN to STARTED to reflect
   the number of started jobs and not only ran.
-- lsid prints the 4.0beta2 identification for the current branch.
-- Implement the decay slot factor for HIST_MINUTES.

4.0beta1
========
-- Fix mbatchd core dump that happens if jobs are submitted with -t option
   of bsub and the job is terminated while pending.
-- Added bjobs -o to support custom output fields
-- Added bjobs -cwd to specify the job current working directory
-- Implement MBD_DEDICATED_RESOURCES in lsf.conf. These are host based boolean
   resources that must be defined in lsf.cluster file. A job must explicitly
   specify a dedicated resource to get allocated the host with that resource.
-- Fix a typo in lsb.queues.in, preemption was spelled with double t.
-- Fix compatibility problems when replaying older lsb.events.
-- Fix the XDR error in the badmin qclose command.
-- Add man page for $LSF_ENVDIR/openlava
-- Add a template file $LSF_ENVDIR/limits.sh
-- Fixed sbatchd compile error on RHEL 5 because glibc lacks CPU_COUNT.
-- Fix a fairshare bug in which jobs were not running under the correct
   user group if bsub -G was specified.
-- Deleted warning message if mbatchd cannot find lsb.resources file.
   lsb.resources is optional.
-- Updated lsb.resources man page for supported use case.
-- Fixed binding incorrect cpus to a queue when HOSTS_SHARES is configured
   in lsb.queues.
-- Enlarge the -J option for the job name to be at least 4096 characters.
-- Added support for binding multiple slots to a core by introducing HOSTS_SHARES
   parameter in lsb.queues for specifying the number of shares the queue is
   allocated for particular host/hostgroup. The share represents the proportion of
   the total number of cores on a host that will be allocated to the queue.
-- Updated non-numa cpu binding feature to support binding multiple
   cores to a job.
-- Added slot limit per host for specific queues in lsb.resources.
-- Added lsb.params PREEMPTABLE_RESOURCES parameters. This enable resource
   preemption in MBD If this parameter is configured the queue preemption
   is done based on the specified resources and not by the host slots which is
   the default.
-- Added the blaunch.1 man page in the batch man section.
-- Developed new API ls_getrpid() to get a remote process id of a task
   started by RES knowing the local taskid.
-- Developed new API lsb_launch() to start a components of parallel jobs.
   This API also collects resource usage of the remote tasks and reports
   the to OpenLava so they can be displayed with the bjobs command.
-- Developed the blaunch command, if mpiexec/mpirun is integrated with
   blaunch, it can launch mpi parallel application, or it can launch
   home grown parallel applications. It is a wrapper around the lsb_launch()
   API.
-- In the bootstrap.sh file create the revision number file
   which is going to be installed in the etc directory.
-- Create lsid 4.0dev4 output.

4.0pre3
=======

-- Create lsid 4.0pre3 tag output.

4.0pre2
========

Introduce 4.0dev3 lsid output. The dev indicates that the code
is between tags. 4.0dev3 finishes with 4.0pre3 tag. After which
we will start 4.0dev4 lsid output till we tag 4.0pre4 or we cut
a final 4.0 branch. In this way when building tagged version we
know the exact commit.

4.0pre1
========

-- Fix res/nios hang blocking ol make.
-- Create a new parameter NIOS_MAX_TASKS specifying the maximum number
   of tasks nios can handle.
-- Fix/reimplement btop/bbot. Since the job queue management has undergone
   performance improvements these function had to be coded as well.
-- Fix the mbatchd variable MBD_NO_QSORT_JOBS, which was incorrectly
   set to MBD_QSORT_JOBS in daemons.c
-- Fix formatting bug in bparams.
-- Add badmin hclose/hopen -C "message", and fix parsing error of "badmin
   subcommand <command_options>".
-- Fix backward compatibility bug in which earlier version of OL cannot
   get information from higher version if mbatchd, specifically 3.0/3.1
   bqueues cannot get information from 3.3 bqueues.
-- Implement lsb.resources of job slot limit for each user/project.
-- Implement the time window in which resource limits are performed.
-- Fix a broken btop/bbot.
-- Introduce a new rule in config/Makefile.am which creates a file named
   commit in the LSF_ENVDIR, this file contains the git commit number
   of the current build.
-- Fix user group bug in lsb.resources.

lsbatch/lib/lsb.err.c
/*
 * Copyright (C) 2014 - 2016 David Bigagli
 * Copyright (C) 2007 Platform Computing Inc
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 *
 */

#include "lsb.h"

int lsberrno = 0;

char   *lsb_errmsg[] = {
    /*00*/   "No error",
    /*01*/   "No matching job found",
    /*02*/   "Job has not started yet",
    /*03*/   "Job has already started",
    /*04*/   "Job has already finished",
    /*05*/   "Error 5", /* no message needed */
    /*06*/   "Dependency condition syntax error",
    /*07*/   "Queue does not accept EXCLUSIVE jobs",
    /*08*/   "Root job submission is disabled",
    /*09*/   "Job is already being migrated",
    /*10*/   "Job is not checkpointable",
    /*11*/   "No output so far",
    /*12*/   "No job Id can be used now",
    /*13*/   "Queue only accepts interactive jobs",
    /*14*/   "Queue does not accept interactive jobs",
    /*15*/   "No user is defined in the lsb.users file",
    /*16*/   "Unknown user",
    /*17*/   "User permission denied",
    /*18*/   "No such queue",
    /*19*/   "Queue name must be specified",
    /*20*/   "Queue has been closed",
    /*21*/   "Not activated because queue windows are closed",
    /*22*/   "User cannot use the queue",
    /*23*/   "Bad host name, host group name or cluster name",
    /*24*/   "Too many processors requested",
    /*25*/   "Reserved for future use",
    /*26*/   "Reserved for future use",
    /*27*/   "No user/host group defined in the system",
    /*28*/   "No such user/host group",
    /*29*/   "Host or host group is not used by the queue",
    /*30*/   "Queue does not have enough per-user job slots",
    /*31*/   "Current host is more suitable at this time",
    /*32*/   "Checkpoint log is not found or is corrupted",
    /*33*/   "Queue does not have enough per-processor job slots",
    /*34*/   "Request from non-LSF host rejected",
    /*35*/   "Bad argument",
    /*36*/   "Bad time specification",
    /*37*/   "Start time is later than termination time",
    /*38*/   "Bad CPU limit specification",
    /*39*/   "Cannot exceed queue's hard limit(s)",
    /*40*/   "Empty job",
    /*41*/   "Signal not supported",
    /*42*/   "Bad job name",
    /*43*/   "The destination queue has reached its job limit",
    /*44*/   "Unknown event",
    /*45*/   "Bad event format",
    /*46*/   "End of file",
    /*47*/   "Master batch daemon internal error",
    /*48*/   "Slave batch daemon internal error",
    /*49*/   "Batch library internal error",
    /*50*/   "Failed in an LSF library call",
    /*51*/   "System call failed",
    /*52*/   "Cannot allocate memory",
    /*53*/   "Batch service not registered",
    /*54*/   "LSB_SHAREDIR not defined",
    /*55*/   "Checkpoint system call failed",
    /*56*/   "Batch daemon cannot fork",
    /*57*/   "Batch protocol error",
    /*58*/   "XDR encode/decode error",
    /*59*/   "Fail to bind to an appropriate port number",
    /*60*/   "Contacting batch daemon: Communication timeout",
    /*61*/   "Timeout on connect call to server",
    /*62*/   "Connection refused by server",
    /*63*/   "Server connection already exists",
    /*64*/   "Server is not connected",
    /*65*/   "Unable to contact execution host",
    /*66*/   "Operation is in progress",
    /*67*/   "User or one of user's groups does not have enough job slots",
    /*68*/   "Job parameters cannot be changed now; non-repetitive job is running",
    /*69*/   "Modified parameters have not been used",
    /*70*/   "Job cannot be run more than once",
    /*71*/   "Unknown cluster name or cluster master",
    /*72*/   "Modified parameters are being used",
    /*73*/   "Queue does not have enough per-host job slots",
    /*74*/   "Mbatchd could not find the message that SBD mentions about",
    /*75*/   "Bad resource requirement syntax",
    /*76*/   "Not enough host(s) currently eligible",
    /*77*/   "Error 77",
    /*78*/   "Error 78",
    /*79*/   "No resource defined",
    /*80*/   "Bad resource name",
    /*81*/   "Interactive job cannot be rerunnable",
    /*82 */   "Input file not allowed with pseudo-terminal",
    /*83*/   "Cannot find restarted or newly submitted job's submission host and host type",
    /*84*/   "Error 109",
    /*85*/   "User not in the specified user group",
    /*86*/   "Cannot exceed queue's resource reservation",
    /*87*/   "Bad host specification",
    /*88*/   "Bad user group name",
    /*89*/   "Request aborted by esub",
    /*90*/   "Bad or invalid action specification",
    /*91*/   "Has dependent jobs",
    /*92*/   "Job group does not exist",
    /*93*/   "Bad/empty job group name",
    /*94*/   "Cannot operate on job array",
    /*95*/   "Operation not supported for a suspended job",
    /*96*/   "Operation not supported for a forwarded job",
    /*97*/   "Job array index error",
    /*98*/   "Job array index too large",
    /*99*/   "Job array does not exist",
    /*100*/   "Job exists",
    /*101*/   "Cannot operate on element job",
    /*102*/   "Bad jobId",
    /*103*/   "Change job name is not allowed for job array",
    /*104*/   "Child process died",
    /*105*/   "Invoker is not in specified project group",
    /*106*/   "No host group defined in the system",
    /*107*/   "No user group defined in the system",
    /*108*/   "Unknown jobid index file format",
    /*109*/   "Source file for spooling does not exist",
    /*110*/   "Number of failed spool hosts reached max",
    /*111*/   "Spool copy failed for this host",
    /*112*/   "Fork for spooling failed",
    /*113*/   "Status of spool child is not available",
    /*114*/   "Spool child terminated with failure",
    /*115*/   "Unable to find a host for spooling",
    /*116*/   "Cannot get $JOB_SPOOL_DIR for this host",
    /*117*/   "Cannot delete spool file for this host",
    /*118*/   "Bad user priority",
    /*119*/   "Job priority control undefined",
    /*120*/   "Job has already been requeued",
    /*121*/   "Multiple first execution hosts specified",
    /*122*/   "Host group specified as first execution host",
    /*123*/   "Host partition specified as first execution host",
    /*124*/   "\"Others\" specified as first execution host",
    /*125*/   "Too few processors requested",
    /*126*/   "Only the following parameters can be used to modify a running job: -c, -M, -W, -o, -e, -r",
    /*127*/   "You must set LSB_JOB_CPULIMIT in lsf.conf to modify the CPU limit of a running job",
    /*128*/   "You must set LSB_JOB_MEMLIMIT in lsf.conf to modify the memory limit of a running job",
    /*129*/   "No error file specified before job dispatch. Error file does not exist, so error file name cannot be changed",
    /*130*/   "The host is locked by master LIM",
    /*131*/  "Dependent arrays do not have the same size",
    /*132*/   "Job group exists already",
    /*133*/   "Job has no dependencies",
    /*134*/   "The job group is not empty",
    /*135*/   "The modification/creation violates the job group limit",
    /* when you add a new message here, remember  to not
     * forget to add "," after the error message otherwise
     * the error count will be wrong.
     */
    NULL
};

char *
lsb_sysmsg(void)
{
    static char buf[512];

    if (lsberrno >= LSBE_NUM_ERR) {
        sprintf(buf, "Unknown batch system error number %d", lsberrno);
        return buf;
    }

    if (lsberrno == LSBE_SYS_CALL) {
        if (errno > 0)
            sprintf(buf, "%s: %s", lsb_errmsg[lsberrno], strerror(errno));
        else
            sprintf(buf, "%s: unknown error %d", lsb_errmsg[lsberrno], errno);
    } else if (lsberrno == LSBE_LSLIB) {
        sprintf(buf, "%s: %s", lsb_errmsg[lsberrno], ls_sysmsg());
    } else {
        return lsb_errmsg[lsberrno];
    }

    return buf;
}

void
lsb_perror(const char *usrMsg)
{
    if (usrMsg) {
        fputs(usrMsg, stderr);
        fputs(": ", stderr);
    }
    fputs(lsb_sysmsg(), stderr);
    putc('\n', stderr);

}

char *
lsb_sperror(char *usrMsg)
{
    char errmsg[256];
    char *rtstr;

    errmsg[0] = '\0';

    if (usrMsg)
        sprintf(errmsg, "%s: ", usrMsg);

    strcat(errmsg, lsb_sysmsg());

    if ((rtstr = malloc(sizeof(char) * (strlen(errmsg) + 1))) == NULL) {
        lserrno = LSE_MALLOC;
        return NULL;
    }

    strcpy(rtstr, errmsg);
    return rtstr;
}



void
sub_perror(char *usrMsg)
{
    if (usrMsg) {
        fputs(usrMsg, stderr);
        fputs(": ", stderr);
    }
    fputs(lsb_sysmsg(), stderr);

}
